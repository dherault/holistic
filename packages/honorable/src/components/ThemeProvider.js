import React from 'react'
import merge from 'lodash.merge'
import { ThemeProvider as EmotionProvider } from '@emotion/react'

import ThemeContext from '../contexts/ThemeContext'

import defaultTheme from '../data/defaultTheme'
import enhanceTheme from '../utils/enhanceTheme'

function ThemeProvider({ theme = {}, children }) {
  const userTheme = enhanceTheme(merge({}, defaultTheme, theme))

  return (
    <ThemeContext.Provider value={userTheme}>
      <EmotionProvider theme={userTheme}>
        {children}
      </EmotionProvider>
    </ThemeContext.Provider>
  )
}

export default ThemeProvider
