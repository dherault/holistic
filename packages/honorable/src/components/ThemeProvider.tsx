import { PropsWithChildren, useContext } from 'react'
import { ThemeProvider as EmotionProvider } from '@emotion/react'
import PropTypes from 'prop-types'

import { ThemeProps } from '../types'

import ThemeContext from '../contexts/ThemeContext'
import defaultTheme from '../data/defaultTheme'
import enhanceTheme from '../utils/enhanceTheme'
import mergeTheme from '../utils/mergeTheme'

export type ThemeProviderProps = PropsWithChildren<ThemeProps>

export const themProviderPropTypes = {
  theme: PropTypes.object,
}

export function ThemeProvider({ theme = {}, children }: ThemeProviderProps) {
  const userTheme = enhanceTheme(mergeTheme(defaultTheme, theme))

  return (
    <ThemeContext.Provider value={userTheme}>
      <EmotionProvider theme={userTheme}>
        {children}
      </EmotionProvider>
    </ThemeContext.Provider>
  )
}

export function ExtendTheme({ theme = {}, children }: ThemeProviderProps) {
  const existingTheme = useContext(ThemeContext)
  const extendedTheme = enhanceTheme(mergeTheme(existingTheme, theme))

  return (
    <ThemeContext.Provider value={extendedTheme}>
      {children}
    </ThemeContext.Provider>
  )
}

ThemeProvider.propTypes = themProviderPropTypes
ExtendTheme.propTypes = themProviderPropTypes
