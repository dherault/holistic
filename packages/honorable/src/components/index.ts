// These are only component that are implicated in the theme
export { default as Button } from './Button'
export { default as ButtonGroup } from './ButtonGroup'
export { default as Caret } from './Caret'
export { default as Checkbox } from './Checkbox'
export { default as DropdownButton } from './DropdownButton'
export { default as Icon } from './Icon'
export { default as IconButton } from './IconButton'
export { default as Menu } from './Menu'
export { default as MenuItem } from './MenuItem'
export { default as Modal } from './Modal'
export { default as ProgressBar } from './ProgressBar'
export { default as Select } from './Select'
export { default as Switch } from './Switch'
export * from './tags'
