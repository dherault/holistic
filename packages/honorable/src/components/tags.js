const { default: wrapComponentWithStyle } = require('../utils/wrapComponentWithStyle')
const { default: capitalize } = require('../utils/capitalize')

module.exports = Object.fromEntries(
  [
    'html',
    'base',
    'head',
    'link',
    'meta',
    'style',
    'title',
    'body',
    'address',
    'article',
    'aside',
    'footer',
    'header',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'main',
    'nav',
    'section',
    'blockquote',
    'dd',
    'div',
    'dl',
    'dt',
    'figcaption',
    'figure',
    'hr',
    'li',
    'menu',
    'ol',
    'p',
    'pre',
    'ul',
    'a',
    'abbr',
    'b',
    'bdi',
    'bdo',
    'br',
    'cite',
    'code',
    'data',
    'dfn',
    'em',
    'i',
    'kbd',
    'mark',
    'q',
    'rp',
    'rt',
    'ruby',
    's',
    'samp',
    'small',
    'span',
    'strong',
    'sub',
    'sup',
    'time',
    'u',
    'var',
    'wbr',
    'area',
    'audio',
    'img',
    'map',
    'track',
    'video',
    'embed',
    'iframe',
    'object',
    'param',
    'picture',
    'portal',
    'source',
    'svg',
    'canvas',
    'noscript',
    'script',
    'del',
    'ins',
    'caption',
    'col',
    'colgroup',
    'table',
    'tbody',
    'td',
    'tfoot',
    'th',
    'thead',
    'tr',
    'button',
    'datalist',
    'fieldset',
    'form',
    'input',
    'label',
    'legend',
    'meter',
    'optgroup',
    'option',
    'output',
    'progress',
    'select',
    'textarea',
    'details',
    'dialog',
    'summary',
    'slot',
    'template',
    'acronym',
    'applet',
    'basefont',
    'bgsound',
    'big',
    'blink',
    'center',
    'content',
    'dir',
    'font',
    'frame',
    'frameset',
    'hgroup',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    'image',
    'keygen',
    'marquee',
    'menuitem',
    'nobr',
    'noembed',
    'noframes',
    'plaintext',
    'rb',
    'rtc',
    'shadow',
    'spacer',
    'strike',
    'tt',
    'xmp',
  ]
.map(tag => [capitalize(tag), wrapComponentWithStyle(tag, tag)])
)
