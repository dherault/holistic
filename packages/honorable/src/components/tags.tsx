import { ElementProps, StylesProps } from '../types'

import withHonorableTag from '../withHonorableTag'

/* --
  * HTML
-- */

export type ButtonBaseBaseProps = StylesProps & ElementProps<'button'>

export const ButtonBase = withHonorableTag<ButtonBaseBaseProps>('button', 'ButtonBase')

export type InputBaseBaseProps = StylesProps & ElementProps<'input'>

export const InputBase = withHonorableTag<InputBaseBaseProps>('input', 'InputBase')

export type MenuBaseBaseProps = StylesProps & ElementProps<'menu'>

export const MenuBase = withHonorableTag<MenuBaseBaseProps>('menu', 'MenuBase')

export type SelectBaseBaseProps = StylesProps & ElementProps<'select'>

export const SelectBase = withHonorableTag<SelectBaseBaseProps>('select', 'SelectBase')

export type AProps = StylesProps & ElementProps<'a'>

export const A = withHonorableTag<AProps>('a', 'A')

export type AbbrProps = StylesProps & ElementProps<'abbr'>

export const Abbr = withHonorableTag<AbbrProps>('abbr', 'Abbr')

export type AddressProps = StylesProps & ElementProps<'address'>

export const Address = withHonorableTag<AddressProps>('address', 'Address')

export type AreaProps = StylesProps & ElementProps<'area'>

export const Area = withHonorableTag<AreaProps>('area', 'Area')

export type ArticleProps = StylesProps & ElementProps<'article'>

export const Article = withHonorableTag<ArticleProps>('article', 'Article')

export type AsideProps = StylesProps & ElementProps<'aside'>

export const Aside = withHonorableTag<AsideProps>('aside', 'Aside')

export type AudioProps = StylesProps & ElementProps<'audio'>

export const Audio = withHonorableTag<AudioProps>('audio', 'Audio')

export type BProps = StylesProps & ElementProps<'b'>

export const B = withHonorableTag<BProps>('b', 'B')

export type BaseProps = StylesProps & ElementProps<'base'>

export const Base = withHonorableTag<BaseProps>('base', 'Base')

export type BdiProps = StylesProps & ElementProps<'bdi'>

export const Bdi = withHonorableTag<BdiProps>('bdi', 'Bdi')

export type BdoProps = StylesProps & ElementProps<'bdo'>

export const Bdo = withHonorableTag<BdoProps>('bdo', 'Bdo')

export type BlockquoteProps = StylesProps & ElementProps<'blockquote'>

export const Blockquote = withHonorableTag<BlockquoteProps>('blockquote', 'Blockquote')

export type BodyProps = StylesProps & ElementProps<'body'>

export const Body = withHonorableTag<BodyProps>('body', 'Body')

export type BrProps = StylesProps & ElementProps<'br'>

export const Br = withHonorableTag<BrProps>('br', 'Br')

export type CanvasProps = StylesProps & ElementProps<'canvas'>

export const Canvas = withHonorableTag<CanvasProps>('canvas', 'Canvas')

export type CaptionProps = StylesProps & ElementProps<'caption'>

export const Caption = withHonorableTag<CaptionProps>('caption', 'Caption')

export type CiteProps = StylesProps & ElementProps<'cite'>

export const Cite = withHonorableTag<CiteProps>('cite', 'Cite')

export type CodeProps = StylesProps & ElementProps<'code'>

export const Code = withHonorableTag<CodeProps>('code', 'Code')

export type ColProps = StylesProps & ElementProps<'col'>

export const Col = withHonorableTag<ColProps>('col', 'Col')

export type ColgroupProps = StylesProps & ElementProps<'colgroup'>

export const Colgroup = withHonorableTag<ColgroupProps>('colgroup', 'Colgroup')

export type DataProps = StylesProps & ElementProps<'data'>

export const Data = withHonorableTag<DataProps>('data', 'Data')

export type DatalistProps = StylesProps & ElementProps<'datalist'>

export const Datalist = withHonorableTag<DatalistProps>('datalist', 'Datalist')

export type DdProps = StylesProps & ElementProps<'dd'>

export const Dd = withHonorableTag<DdProps>('dd', 'Dd')

export type DelProps = StylesProps & ElementProps<'del'>

export const Del = withHonorableTag<DelProps>('del', 'Del')

export type DetailsProps = StylesProps & ElementProps<'details'>

export const Details = withHonorableTag<DetailsProps>('details', 'Details')

export type DfnProps = StylesProps & ElementProps<'dfn'>

export const Dfn = withHonorableTag<DfnProps>('dfn', 'Dfn')

export type DialogProps = StylesProps & ElementProps<'dialog'>

export const Dialog = withHonorableTag<DialogProps>('dialog', 'Dialog')

export type DivProps = StylesProps & ElementProps<'div'>

export const Div = withHonorableTag<DivProps>('div', 'Div')

export type DlProps = StylesProps & ElementProps<'dl'>

export const Dl = withHonorableTag<DlProps>('dl', 'Dl')

export type DtProps = StylesProps & ElementProps<'dt'>

export const Dt = withHonorableTag<DtProps>('dt', 'Dt')

export type EmProps = StylesProps & ElementProps<'em'>

export const Em = withHonorableTag<EmProps>('em', 'Em')

export type EmbedProps = StylesProps & ElementProps<'embed'>

export const Embed = withHonorableTag<EmbedProps>('embed', 'Embed')

export type FieldsetProps = StylesProps & ElementProps<'fieldset'>

export const Fieldset = withHonorableTag<FieldsetProps>('fieldset', 'Fieldset')

export type FigcaptionProps = StylesProps & ElementProps<'figcaption'>

export const Figcaption = withHonorableTag<FigcaptionProps>('figcaption', 'Figcaption')

export type FigureProps = StylesProps & ElementProps<'figure'>

export const Figure = withHonorableTag<FigureProps>('figure', 'Figure')

export type FooterProps = StylesProps & ElementProps<'footer'>

export const Footer = withHonorableTag<FooterProps>('footer', 'Footer')

export type FormProps = StylesProps & ElementProps<'form'>

export const Form = withHonorableTag<FormProps>('form', 'Form')

export type H1Props = StylesProps & ElementProps<'h1'>

export const H1 = withHonorableTag<H1Props>('h1', 'H1')

export type H2Props = StylesProps & ElementProps<'h2'>

export const H2 = withHonorableTag<H2Props>('h2', 'H2')

export type H3Props = StylesProps & ElementProps<'h3'>

export const H3 = withHonorableTag<H3Props>('h3', 'H3')

export type H4Props = StylesProps & ElementProps<'h4'>

export const H4 = withHonorableTag<H4Props>('h4', 'H4')

export type H5Props = StylesProps & ElementProps<'h5'>

export const H5 = withHonorableTag<H5Props>('h5', 'H5')

export type H6Props = StylesProps & ElementProps<'h6'>

export const H6 = withHonorableTag<H6Props>('h6', 'H6')

export type HeadProps = StylesProps & ElementProps<'head'>

export const Head = withHonorableTag<HeadProps>('head', 'Head')

export type HeaderProps = StylesProps & ElementProps<'header'>

export const Header = withHonorableTag<HeaderProps>('header', 'Header')

export type HrProps = StylesProps & ElementProps<'hr'>

export const Hr = withHonorableTag<HrProps>('hr', 'Hr')

// export type HtmlProps = StylesProps & ElementProps<'html'>
// export const Html = withHonorableTag<HtmlProps>('html', 'html') // Can't use this because it's a reserved word in the Theme
export type IProps = StylesProps & ElementProps<'i'>

export const I = withHonorableTag<IProps>('i', 'I')

export type IframeProps = StylesProps & ElementProps<'iframe'>

export const Iframe = withHonorableTag<IframeProps>('iframe', 'Iframe')

export type ImgProps = StylesProps & ElementProps<'img'>

export const Img = withHonorableTag<ImgProps>('img', 'Img')

export type InsProps = StylesProps & ElementProps<'ins'>

export const Ins = withHonorableTag<InsProps>('ins', 'Ins')

export type KbdProps = StylesProps & ElementProps<'kbd'>

export const Kbd = withHonorableTag<KbdProps>('kbd', 'Kbd')

export type LabelProps = StylesProps & ElementProps<'label'>

export const Label = withHonorableTag<LabelProps>('label', 'Label')

export type LegendProps = StylesProps & ElementProps<'legend'>

export const Legend = withHonorableTag<LegendProps>('legend', 'Legend')

export type LiProps = StylesProps & ElementProps<'li'>

export const Li = withHonorableTag<LiProps>('li', 'Li')

export type LinkProps = StylesProps & ElementProps<'link'>

export const Link = withHonorableTag<LinkProps>('link', 'Link')

export type MainProps = StylesProps & ElementProps<'main'>

export const Main = withHonorableTag<MainProps>('main', 'Main')

export type MapProps = StylesProps & ElementProps<'map'>

export const Map = withHonorableTag<MapProps>('map', 'Map')

export type MarkProps = StylesProps & ElementProps<'mark'>

export const Mark = withHonorableTag<MarkProps>('mark', 'Mark')

export type MetaProps = StylesProps & ElementProps<'meta'>

export const Meta = withHonorableTag<MetaProps>('meta', 'Meta')

export type MeterProps = StylesProps & ElementProps<'meter'>

export const Meter = withHonorableTag<MeterProps>('meter', 'Meter')

export type NavProps = StylesProps & ElementProps<'nav'>

export const Nav = withHonorableTag<NavProps>('nav', 'Nav')

export type NoscriptProps = StylesProps & ElementProps<'noscript'>

export const Noscript = withHonorableTag<NoscriptProps>('noscript', 'Noscript')

// export type ObjectProps = StylesProps & ElementProps<'object'>
// export const Object = withHonorableTag<ObjectProps>('object', 'object') // Can't use this because it's a reserved Word
export type OlProps = StylesProps & ElementProps<'ol'>

export const Ol = withHonorableTag<OlProps>('ol', 'Ol')

export type OptgroupProps = StylesProps & ElementProps<'optgroup'>

export const Optgroup = withHonorableTag<OptgroupProps>('optgroup', 'Optgroup')

export type OptionProps = StylesProps & ElementProps<'option'>

export const Option = withHonorableTag<OptionProps>('option', 'Option')

export type OutputProps = StylesProps & ElementProps<'output'>

export const Output = withHonorableTag<OutputProps>('output', 'Output')

export type PProps = StylesProps & ElementProps<'p'>

export const P = withHonorableTag<PProps>('p', 'P')

export type ParamProps = StylesProps & ElementProps<'param'>

export const Param = withHonorableTag<ParamProps>('param', 'Param')

export type PictureProps = StylesProps & ElementProps<'picture'>

export const Picture = withHonorableTag<PictureProps>('picture', 'Picture')

// export type PortalProps = StylesProps & ElementProps<'portal'>
// export const Portal = withHonorableTag<PortalProps>('portal', 'Portal')
export type PreProps = StylesProps & ElementProps<'pre'>

export const Pre = withHonorableTag<PreProps>('pre', 'Pre')

export type ProgressProps = StylesProps & ElementProps<'progress'>

export const Progress = withHonorableTag<ProgressProps>('progress', 'Progress')

export type QProps = StylesProps & ElementProps<'q'>

export const Q = withHonorableTag<QProps>('q', 'Q')

export type RpProps = StylesProps & ElementProps<'rp'>

export const Rp = withHonorableTag<RpProps>('rp', 'Rp')

export type RtProps = StylesProps & ElementProps<'rt'>

export const Rt = withHonorableTag<RtProps>('rt', 'Rt')

export type RubyProps = StylesProps & ElementProps<'ruby'>

export const Ruby = withHonorableTag<RubyProps>('ruby', 'Ruby')

export type SProps = StylesProps & ElementProps<'s'>

export const S = withHonorableTag<SProps>('s', 'S')

export type SampProps = StylesProps & ElementProps<'samp'>

export const Samp = withHonorableTag<SampProps>('samp', 'Samp')

export type ScriptProps = StylesProps & ElementProps<'script'>

export const Script = withHonorableTag<ScriptProps>('script', 'Script')

export type SectionProps = StylesProps & ElementProps<'section'>

export const Section = withHonorableTag<SectionProps>('section', 'Section')

export type SlotProps = StylesProps & ElementProps<'slot'>

export const Slot = withHonorableTag<SlotProps>('slot', 'Slot')

export type SmallProps = StylesProps & ElementProps<'small'>

export const Small = withHonorableTag<SmallProps>('small', 'Small')

export type SourceProps = StylesProps & ElementProps<'source'>

export const Source = withHonorableTag<SourceProps>('source', 'Source')

export type SpanProps = StylesProps & ElementProps<'span'>

export const Span = withHonorableTag<SpanProps>('span', 'Span')

export type StrongProps = StylesProps & ElementProps<'strong'>

export const Strong = withHonorableTag<StrongProps>('strong', 'Strong')

export type StyleProps = StylesProps & ElementProps<'style'>

export const Style = withHonorableTag<StyleProps>('style', 'Style')

export type SubProps = StylesProps & ElementProps<'sub'>

export const Sub = withHonorableTag<SubProps>('sub', 'Sub')

export type SummaryProps = StylesProps & ElementProps<'summary'>

export const Summary = withHonorableTag<SummaryProps>('summary', 'Summary')

export type SupProps = StylesProps & ElementProps<'sup'>

export const Sup = withHonorableTag<SupProps>('sup', 'Sup')

export type SvgProps = StylesProps & ElementProps<'svg'>

export const Svg = withHonorableTag<SvgProps>('svg', 'Svg')

export type TableProps = StylesProps & ElementProps<'table'>

export const Table = withHonorableTag<TableProps>('table', 'Table')

export type TbodyProps = StylesProps & ElementProps<'tbody'>

export const Tbody = withHonorableTag<TbodyProps>('tbody', 'Tbody')

export type TdProps = StylesProps & ElementProps<'td'>

export const Td = withHonorableTag<TdProps>('td', 'Td')

export type TemplateProps = StylesProps & ElementProps<'template'>

export const Template = withHonorableTag<TemplateProps>('template', 'Template')

export type TextareaProps = StylesProps & ElementProps<'textarea'>

export const Textarea = withHonorableTag<TextareaProps>('textarea', 'Textarea')

export type TfootProps = StylesProps & ElementProps<'tfoot'>

export const Tfoot = withHonorableTag<TfootProps>('tfoot', 'Tfoot')

export type ThProps = StylesProps & ElementProps<'th'>

export const Th = withHonorableTag<ThProps>('th', 'Th')

export type TheadProps = StylesProps & ElementProps<'thead'>

export const Thead = withHonorableTag<TheadProps>('thead', 'Thead')

export type TimeProps = StylesProps & ElementProps<'time'>

export const Time = withHonorableTag<TimeProps>('time', 'Time')

export type TitleProps = StylesProps & ElementProps<'title'>

export const Title = withHonorableTag<TitleProps>('title', 'Title')

export type TrProps = StylesProps & ElementProps<'tr'>

export const Tr = withHonorableTag<TrProps>('tr', 'Tr')

export type TrackProps = StylesProps & ElementProps<'track'>

export const Track = withHonorableTag<TrackProps>('track', 'Track')

export type UProps = StylesProps & ElementProps<'u'>

export const U = withHonorableTag<UProps>('u', 'U')

export type UlProps = StylesProps & ElementProps<'ul'>

export const Ul = withHonorableTag<UlProps>('ul', 'Ul')

export type VarProps = StylesProps & ElementProps<'var'>

export const Var = withHonorableTag<VarProps>('var', 'Var')

export type VideoProps = StylesProps & ElementProps<'video'>

export const Video = withHonorableTag<VideoProps>('video', 'Video')

export type WbrProps = StylesProps & ElementProps<'wbr'>

export const Wbr = withHonorableTag<WbrProps>('wbr', 'Wbr')

/* --
  * SVG
-- */

export type AnimateProps = StyleProps & ElementProps<'animate'>

export const Animate = withHonorableTag<AnimateProps>('animate', 'Animate')

export type AnimateMotionProps = StyleProps & ElementProps<'animateMotion'>

export const AnimateMotion = withHonorableTag<AnimateMotionProps>('animateMotion', 'AnimateMotion')

export type AnimateTransformProps = StyleProps & ElementProps<'animateTransform'>

export const AnimateTransform = withHonorableTag<AnimateTransformProps>('animateTransform', 'AnimateTransform')

export type CircleProps = StyleProps & ElementProps<'circle'>

export const Circle = withHonorableTag<CircleProps>('circle', 'Circle')

export type ClipPathProps = StyleProps & ElementProps<'clipPath'>

export const ClipPath = withHonorableTag<ClipPathProps>('clipPath', 'ClipPath')

export type DefsProps = StyleProps & ElementProps<'defs'>

export const Defs = withHonorableTag<DefsProps>('defs', 'Defs')

export type DescProps = StyleProps & ElementProps<'desc'>

export const Desc = withHonorableTag<DescProps>('desc', 'Desc')

export type EllipseProps = StyleProps & ElementProps<'ellipse'>

export const Ellipse = withHonorableTag<EllipseProps>('ellipse', 'Ellipse')

export type FeBlendProps = StyleProps & ElementProps<'feBlend'>

export const FeBlend = withHonorableTag<FeBlendProps>('feBlend', 'FeBlend')

export type FeColorMatrixProps = StyleProps & ElementProps<'feColorMatrix'>

export const FeColorMatrix = withHonorableTag<FeColorMatrixProps>('feColorMatrix', 'FeColorMatrix')

export type FeComponentTransferProps = StyleProps & ElementProps<'feComponentTransfer'>

export const FeComponentTransfer = withHonorableTag<FeComponentTransferProps>('feComponentTransfer', 'FeComponentTransfer')

export type FeCompositeProps = StyleProps & ElementProps<'feComposite'>

export const FeComposite = withHonorableTag<FeCompositeProps>('feComposite', 'FeComposite')

export type FeConvolveMatrixProps = StyleProps & ElementProps<'feConvolveMatrix'>

export const FeConvolveMatrix = withHonorableTag<FeConvolveMatrixProps>('feConvolveMatrix', 'FeConvolveMatrix')

export type FeDiffuseLightingProps = StyleProps & ElementProps<'feDiffuseLighting'>

export const FeDiffuseLighting = withHonorableTag<FeDiffuseLightingProps>('feDiffuseLighting', 'FeDiffuseLighting')

export type FeDisplacementMapProps = StyleProps & ElementProps<'feDisplacementMap'>

export const FeDisplacementMap = withHonorableTag<FeDisplacementMapProps>('feDisplacementMap', 'FeDisplacementMap')

export type FeDistantLightProps = StyleProps & ElementProps<'feDistantLight'>

export const FeDistantLight = withHonorableTag<FeDistantLightProps>('feDistantLight', 'FeDistantLight')

export type FeDropShadowProps = StyleProps & ElementProps<'feDropShadow'>

export const FeDropShadow = withHonorableTag<FeDropShadowProps>('feDropShadow', 'FeDropShadow')

export type FeFloodProps = StyleProps & ElementProps<'feFlood'>

export const FeFlood = withHonorableTag<FeFloodProps>('feFlood', 'FeFlood')

export type FeFuncAProps = StyleProps & ElementProps<'feFuncA'>

export const FeFuncA = withHonorableTag<FeFuncAProps>('feFuncA', 'FeFuncA')

export type FeFuncBProps = StyleProps & ElementProps<'feFuncB'>

export const FeFuncB = withHonorableTag<FeFuncBProps>('feFuncB', 'FeFuncB')

export type FeFuncGProps = StyleProps & ElementProps<'feFuncG'>

export const FeFuncG = withHonorableTag<FeFuncGProps>('feFuncG', 'FeFuncG')

export type FeFuncRProps = StyleProps & ElementProps<'feFuncR'>

export const FeFuncR = withHonorableTag<FeFuncRProps>('feFuncR', 'FeFuncR')

export type FeGaussianBlurProps = StyleProps & ElementProps<'feGaussianBlur'>

export const FeGaussianBlur = withHonorableTag<FeGaussianBlurProps>('feGaussianBlur', 'FeGaussianBlur')

export type FeImageProps = StyleProps & ElementProps<'feImage'>

export const FeImage = withHonorableTag<FeImageProps>('feImage', 'FeImage')

export type FeMergeProps = StyleProps & ElementProps<'feMerge'>

export const FeMerge = withHonorableTag<FeMergeProps>('feMerge', 'FeMerge')

export type FeMergeNodeProps = StyleProps & ElementProps<'feMergeNode'>

export const FeMergeNode = withHonorableTag<FeMergeNodeProps>('feMergeNode', 'FeMergeNode')

export type FeMorphologyProps = StyleProps & ElementProps<'feMorphology'>

export const FeMorphology = withHonorableTag<FeMorphologyProps>('feMorphology', 'FeMorphology')

export type FeOffsetProps = StyleProps & ElementProps<'feOffset'>

export const FeOffset = withHonorableTag<FeOffsetProps>('feOffset', 'FeOffset')

export type FePointLightProps = StyleProps & ElementProps<'fePointLight'>

export const FePointLight = withHonorableTag<FePointLightProps>('fePointLight', 'FePointLight')

export type FeSpecularLightingProps = StyleProps & ElementProps<'feSpecularLighting'>

export const FeSpecularLighting = withHonorableTag<FeSpecularLightingProps>('feSpecularLighting', 'FeSpecularLighting')

export type FeSpotLightProps = StyleProps & ElementProps<'feSpotLight'>

export const FeSpotLight = withHonorableTag<FeSpotLightProps>('feSpotLight', 'FeSpotLight')

export type FeTileProps = StyleProps & ElementProps<'feTile'>

export const FeTile = withHonorableTag<FeTileProps>('feTile', 'FeTile')

export type FeTurbulenceProps = StyleProps & ElementProps<'feTurbulence'>

export const FeTurbulence = withHonorableTag<FeTurbulenceProps>('feTurbulence', 'FeTurbulence')

export type FilterProps = StyleProps & ElementProps<'filter'>

export const Filter = withHonorableTag<FilterProps>('filter', 'Filter')

export type ForeignObjectProps = StyleProps & ElementProps<'foreignObject'>

export const ForeignObject = withHonorableTag<ForeignObjectProps>('foreignObject', 'ForeignObject')

export type GProps = StyleProps & ElementProps<'g'>

export const G = withHonorableTag<GProps>('g', 'G')

export type ImageProps = StyleProps & ElementProps<'image'>

export const Image = withHonorableTag<ImageProps>('image', 'Image')

export type LineProps = StyleProps & ElementProps<'line'>

export const Line = withHonorableTag<LineProps>('line', 'Line')

export type LinearGradientProps = StyleProps & ElementProps<'linearGradient'>

export const LinearGradient = withHonorableTag<LinearGradientProps>('linearGradient', 'LinearGradient')

export type MarkerProps = StyleProps & ElementProps<'marker'>

export const Marker = withHonorableTag<MarkerProps>('marker', 'Marker')

export type MaskProps = StyleProps & ElementProps<'mask'>

export const Mask = withHonorableTag<MaskProps>('mask', 'Mask')

export type MetadataProps = StyleProps & ElementProps<'metadata'>

export const Metadata = withHonorableTag<MetadataProps>('metadata', 'Metadata')

export type MpathProps = StyleProps & ElementProps<'mpath'>

export const Mpath = withHonorableTag<MpathProps>('mpath', 'Mpath')

export type PathProps = StyleProps & ElementProps<'path'>

export const Path = withHonorableTag<PathProps>('path', 'Path')

export type PatternProps = StyleProps & ElementProps<'pattern'>

export const Pattern = withHonorableTag<PatternProps>('pattern', 'Pattern')

export type PolygonProps = StyleProps & ElementProps<'polygon'>

export const Polygon = withHonorableTag<PolygonProps>('polygon', 'Polygon')

export type PolylineProps = StyleProps & ElementProps<'polyline'>

export const Polyline = withHonorableTag<PolylineProps>('polyline', 'Polyline')

export type RadialGradientProps = StyleProps & ElementProps<'radialGradient'>

export const RadialGradient = withHonorableTag<RadialGradientProps>('radialGradient', 'RadialGradient')

export type RectProps = StyleProps & ElementProps<'rect'>

export const Rect = withHonorableTag<RectProps>('rect', 'Rect')

export type StopProps = StyleProps & ElementProps<'stop'>

export const Stop = withHonorableTag<StopProps>('stop', 'Stop')

export type SvgSwitchProps = StyleProps & ElementProps<'switch'>

export const SvgSwitch = withHonorableTag<SvgSwitchProps>('switch', 'Switch')

export type SymbolProps = StyleProps & ElementProps<'symbol'>

export const Symbol = withHonorableTag<SymbolProps>('symbol', 'Symbol')

export type SvgTextProps = StyleProps & ElementProps<'text'>

export const SvgText = withHonorableTag<SvgTextProps>('text', 'Text')

export type TextPathProps = StyleProps & ElementProps<'textPath'>

export const TextPath = withHonorableTag<TextPathProps>('textPath', 'TextPath')

export type TspanProps = StyleProps & ElementProps<'tspan'>

export const Tspan = withHonorableTag<TspanProps>('tspan', 'Tspan')

export type UseProps = StyleProps & ElementProps<'use'>

export const Use = withHonorableTag<UseProps>('use', 'Use')

export type ViewProps = StyleProps & ElementProps<'view'>

export const View = withHonorableTag<ViewProps>('view', 'View')
