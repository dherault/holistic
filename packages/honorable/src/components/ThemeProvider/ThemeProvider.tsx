import { PropsWithChildren, useContext } from 'react'
import { ThemeProvider as EmotionProvider } from '@emotion/react'
import PropTypes from 'prop-types'

import { HonorableTheme } from '../../types'

import ThemeContext from '../../contexts/ThemeContext'
import enhanceTheme from '../../utils/enhanceTheme'
import mergeTheme from '../../utils/mergeTheme'

import { Div } from '../tags'

export type ThemeProviderBaseProps = {
  theme: HonorableTheme
}

export type ThemeProviderProps = PropsWithChildren<ThemeProviderBaseProps>

export const themProviderPropTypes = {
  theme: PropTypes.object,
}

export function ThemeProvider({ theme = {}, children }: ThemeProviderProps) {
  const userTheme = enhanceTheme(theme)

  return (
    <ThemeContext.Provider value={userTheme}>
      <EmotionProvider theme={userTheme}>
        {children}
        <Div id="honorable-portal" />
      </EmotionProvider>
    </ThemeContext.Provider>
  )
}

export function ExtendTheme({ theme = {}, children }: ThemeProviderProps) {
  const existingTheme = useContext(ThemeContext)
  const extendedTheme = enhanceTheme(mergeTheme(existingTheme, theme))

  return (
    <ThemeContext.Provider value={extendedTheme}>
      {children}
    </ThemeContext.Provider>
  )
}

ThemeProvider.propTypes = themProviderPropTypes
ExtendTheme.propTypes = themProviderPropTypes
