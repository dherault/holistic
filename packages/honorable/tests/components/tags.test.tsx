import renderer from 'react-test-renderer'

import '../test-utils'
import theme from '../themes/theme1'

import { Div } from '../../src/components/tags'
import { ThemeProvider } from '../../src/components/ThemeProvider/ThemeProvider'

describe('Div', () => {

  it('Renders as a <div />', () => {
    const component = renderer.create(
      <Div>Honorable</Div>,
    )

    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Renders as a <div /> with styles', () => {
    const component = renderer.create(
      <Div color="blue">Honorable</Div>,
    )

    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('color', 'blue')
  })

  it('Renders as a <div /> with styles on breakpoints', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Div color-mobile="blue">Honorable</Div>
      </ThemeProvider>
    )

    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('color', 'blue', {
      media: '(min-width: 0px) and (max-width: 600px)',
    })
  })

  it('Renders as a <div /> with styles on theme colors', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Div color="primary">Honorable</Div>
      </ThemeProvider>
    )

    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('color', '#0000ff')
  })
})
