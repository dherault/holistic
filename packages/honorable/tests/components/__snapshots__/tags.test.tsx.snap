// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Div Renders as a <div /> 1`] = `
<div
  className="emotion-0"
>
  Honorable
</div>
`;

exports[`Div Renders as a <div /> with styles 1`] = `
.emotion-0 {
  color: blue;
}

<div
  className="emotion-0"
>
  Honorable
</div>
`;

exports[`Div Renders as a <div /> with styles on breakpoints 1`] = `
@media (min-width: 0px) and (max-width: 600px) {
  .emotion-0 {
    color: blue;
  }
}

<div
  className="emotion-0"
>
  Honorable
</div>
`;

exports[`Div Renders as a <div /> with styles on theme colors 1`] = `
.emotion-0 {
  color: #0000ff;
}

<div
  className="emotion-0"
>
  Honorable
</div>
`;
